File: Stacks/20-valid-parentheses.java
ublic class Solution {
    public boolean isValid(String s) {
        char[] ch = s.toCharArray();
        Stack<Character> stack = new Stack<Character>();
        HashMap d = new HashMap();
        d.put(')', '(');
        d.put(']', '[');
        d.put('}', '{');
        
        for(char e: ch){
            if (e=='{' || e=='[' || e=='('){
                stack.push(e);
            }else{
                if (stack.isEmpty()){
                    return false;
                }
                if (d.get(e) == stack.peek()){
                    stack.pop();
                }else{
                    return false;
                }
                
            }
        }
        if (stack.isEmpty()){
            return true;
        }else{
            return false;
        }
        
    }
}

File: Stacks/20-valid-parentheses.py
"""
https://leetcode.com/problems/valid-parentheses/
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
"""

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        d = {")": "(", "]": "[", "}": "{"}
        stack = []
        s = list(s)
        for e in s:
            if e in ("{", "(", "["):
                stack.append(e)
            else:
                if not stack:
                    return False
                if d[e] == stack[-1]:
                    stack.pop()
                else:
                    return False
        if stack:
            return False
        else:
            return True
